lib/chef/provider/package/homebrew.rb

```diff
@@ -24,56 +24,59 @@ class Chef
   class Provider
     class Package
       class Homebrew < Chef::Provider::Package
+        allow_nils
+        use_multipackage_api
 
         provides :package, os: "darwin", override: true
         provides :homebrew_package
 
         include Chef::Mixin::HomebrewUser
 
         def load_current_resource
-          self.current_resource = Chef::Resource::HomebrewPackage.new(new_resource.name)
+          @current_resource = Chef::Resource::HomebrewPackage.new(new_resource.name)
           current_resource.package_name(new_resource.package_name)
-          current_resource.version(current_installed_version)
-          logger.trace("#{new_resource} current version is #{current_resource.version}") if current_resource.version
-
-          @candidate_version = candidate_version
-
-          logger.trace("#{new_resource} candidate version is #{@candidate_version}") if @candidate_version
+          current_resource.version(get_current_versions)
+          logger.trace("#{new_resource} current package version(s): #{current_resource.version}") if current_resource.version
 
           current_resource
         end
 
-        def install_package(name, version)
-          unless current_resource.version == version
-            brew("install", options, name)
+        def candidate_version
+          package_name_array.map do |package_name|
+            available_version(package_name)
           end
         end
 
-        def upgrade_package(name, version)
-          current_version = current_resource.version
-
-          if current_version.nil? || current_version.empty?
-            install_package(name, version)
-          elsif current_version != version
-            brew("upgrade", options, name)
+        def get_current_versions
+          package_name_array.map do |package_name|
+            installed_version(package_name)
           end
         end
 
-        def remove_package(name, version)
-          if current_resource.version
-            brew("uninstall", options, name)
-          end
+        def install_package(names, versions)
+          brew_cmd_output("install", options, names.compact)
         end
 
-        # Homebrew doesn't really have a notion of purging, do a "force remove"
-        def purge_package(name, version)
-          if current_resource.version
-            brew("uninstall", "--force", options, name)
-          end
+        # upgrades are a bit harder in homebrew than other package formats. If you try to
+        # brew upgrade a package that isn't installed it will fail so if a user specifies
+        # the action of upgrade we need to figure out which packages need to be installed
+        # and which packages can be upgrades. We do this by checking if brew_info has an entry
+        # via the installed_version helper.
+        def upgrade_package(names, versions)
+          upgrade_pkgs = names.filter_map { |x| x if installed_version(x) }
+          install_pkgs = names.filter_map { |x| x unless installed_version(x) }
```

Comment:
```suggestion
          upgrade_pkgs, install_pkgs = names.partition { |x| !!installed_version(x) }
```

Reply:
I didn't know `filter_map` was a thing. Neat! ✨

Reply:
Sadly it's not a thing in 2.6